### PR TITLE
Feature/datamodel&mappers

### PR DESCRIPTION
**Pull request by agabylicka**
Finished part datamodel&mappers.

### CODE REVIEW

File: src/main/java/org/example/ARaceByIndex.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package org.example;
+
+import java.util.List;
+import java.util.Objects;
+
+public class ARaceByIndex {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/ARaceByIndex.java:6]: To be deleted

File: src/main/java/org/example/Character.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package org.example;
+
+import java.util.List;
+
+public class Character {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/Character.java:5]: To be deleted

File: src/main/java/org/example/ClazzMAPPER.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ClazzMAPPER {
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    public static Clazz convertJSON(String postJSON) {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/ClazzMAPPER.java]: JSON does not contain any posts

File: src/main/java/org/example/Clazz.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package org.example;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.util.List;
+
+public class Clazz {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/Clazz.java]: Suggested change of name to CharacterClass

File: src/main/java/org/example/Feature.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+public class Feature {
+    private String name;
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    //build JSON
+    @Override
+    public String toString() {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/Feature.java:16]: Please put toString on the bottom

File: src/main/java/org/example/FeatureMAPPER.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package org.example;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class FeatureMAPPER {
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    public static Feature convertJSON(String postJSON) {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/FeatureMAPPER.java]: JSON does not contain post

File: src/main/java/org/example/ItemMAPPER.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package org.example;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class ItemMAPPER {
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    public static Item convertJSON(String postJSON) {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/ItemMAPPER.java]: same with post

File: src/main/java/org/example/Language.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package org.example;
+
+public class Language {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/Language.java:3]: Same as character and class

File: src/main/java/org/example/RaceMAPPER.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+public class RaceMAPPER {
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    public static Race convertJSON(String postJSON) {
+        try {
+            Race race = new Race();
+            JsonNode node = MAPPER.readTree(postJSON);
+            race.setSize((node.get("size").asText()));
+            race.setSpeed((node.get("speed").asInt())); //?
```
Comment (blazejprzyluski) [inline src/main/java/org/example/RaceMAPPER.java]: comment to be deleted

File: src/main/java/org/example/RaceMAPPER.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+            for (int i = 0; i < node.get("ability_bonuses").size(); i++) {
+                AbilityBonus bonus;
+                String tmpBonus = node.get("ability_bonuses").get(i).get("ability_score").get("name").asText();
+                int bonusValue = node.get("ability_bonuses").get(i).get("bonus").asInt();
+                AbilityBonus abilityBonus1 = new AbilityBonus();
+                abilityBonus1.setName(tmpBonus);
+                abilityBonus1.setValue(bonusValue);
+                bonuses.add(abilityBonus1);
+            }
+            return race;
```
Comment (blazejprzyluski) [inline src/main/java/org/example/RaceMAPPER.java:43]: please add all created lists to race object

File: src/main/java/org/example/SpellsMAPPER.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package org.example;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class SpellsMAPPER {
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    public static Spells convertJSON(String postJSON) {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/SpellsMAPPER.java]: same with post

File: src/main/java/org/example/Subrace.java [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package org.example;
+
+public class Subrace {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/Subrace.java:3]: Same as character and class

File: src/main/java/org/example/Trait.java [hunk 13]
```diff
@@ -0,0 +1,9 @@
+package org.example;
+
+public class Trait {
```
Comment (blazejprzyluski) [inline src/main/java/org/example/Trait.java:3]: Same as character and class

File: src/main/java/org/example/race_to_parse.json
Comment (blazejprzyluski) [inline src/main/java/org/example/race_to_parse.json:1]: Please move this to test directory

File: src/test/java/TestAbilityBonus.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+public class TestAbilityBonus {
+    @Test
+    public void testConvertJSONAbilityBonus() {
+        //given
+        String name = "Charisma";
+        String json = "{\"name\": \"" + name + "\",\"value\": \"" + 1 + "\"}";
+        //when
+        AbilityBonus actual = AbilityBonusMAPPER.convertJSON(json);
+        //then
+        Assertions.assertEquals(name, actual.getName());
```
Comment (blazejprzyluski) [inline src/test/java/TestAbilityBonus.java:17]: Lacking check for value

File: src/test/java/TestAbilityBonus.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    @Test
+    public void testConvertJSONAbilityBonusWithWhiteSigns() {
+        //given
+        String name = " Charisma  ";
```
Comment (blazejprzyluski) [inline src/test/java/TestAbilityBonus.java:23]: This is not a strong enough test. Your code will accept everything under the "name" field. Please test cases, where both name and value fields have errors in their field names

File: src/test/java/TestAbilityBonus.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        //given
+        int value = 0;
+        String name = "Charisma";
+        String json = "{\"name\": \"" + name + "\",\"value\": \"" + value + "\"}";
+        //when
+        AbilityBonus actual = AbilityBonusMAPPER.convertJSON(json);
+        //then
+        Assertions.assertEquals(value, actual.getValue());
+    }
+
```
Comment (blazejprzyluski) [inline src/test/java/TestAbilityBonus.java]: To be deleted, refer to the comment above

File: src/test/java/TestClass.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TestClass {
+
+    @Test
+    public void shootPassStartingEquipmentTest() {
```
Comment (blazejprzyluski) [inline src/test/java/TestClass.java]: Please fix test naming

File: src/test/java/TestClass.java [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+        List<Item> testList = new ArrayList<>();
+        Clazz instance = new Clazz();
+
+        //when
+        instance.setStartingEquipment(testList);
+        List<Item> actual = instance.getStartingEquipment();
+
+        //then
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(testList, actual);
```
Comment (blazejprzyluski) [inline src/test/java/TestClass.java:26]: You're testing if two empty lists are empty

File: src/test/java/TestClass.java [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+
+        //when
+        availableSkills.setAvailableSkills((testAvailableSkills));
+        List<String> actual = availableSkills.getAvailableSkills();
+
+        //then
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(testAvailableSkills, actual);
+    }
+
```
Comment (blazejprzyluski) [inline src/test/java/TestClass.java]: same as above

File: src/test/java/TestClass.java [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+        String json= "{ " +
+                "\"proficiencies\" : [\"" + proficiency + "\"]," +
+                "\"available_skills\": [], " +
+                "\"starting_equipment\": []}";
+
+        //when
+        Clazz actual = ClazzMAPPER.convertJSON(json);
+
+        //then
+        Assertions.assertNotNull(actual);
```
Comment (blazejprzyluski) [inline src/test/java/TestClass.java:74]: it's better to compare two lists

File: src/test/java/TestItem.java [hunk 21]
```diff
@@ -0,0 +1,128 @@
+import org.example.Item;
+import org.example.ItemMAPPER;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class TestItem {
```
Comment (blazejprzyluski) [inline src/test/java/TestItem.java:6]: Same feedback as ability bonus

File: src/test/java/TestRace.java [hunk 22]
```diff
@@ -0,0 +1,544 @@
...
+                "      \"url\": \"/api/traits/damage-resistance\"\n" +
+                "    }\n" +
+                "  ],\n" +
+                "  \"subraces\": [],\n" +
+                "  \"url\": \"/api/races/dragonborn\"\n" +
+                "}";
+        //when
+        Race actual = RaceMAPPER.convertJSON(json);
+        //then
+        Assertions.assertEquals(size, actual.getSize());
```
Comment (blazejprzyluski) [inline src/test/java/TestRace.java:77]: assert on all fields

File: src/test/java/TestSpells.java [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+        //then
+        Assertions.assertNull(actual);
+    }
+
+    @Test
+    public void testErroneusConvertJSONSpellsWithWrongNameType() {
+        //given
+        String json = "{\"nname\": \"illusion\"}";
+        Exception thrown = null;
+        //when
```
Comment (blazejprzyluski) [inline src/test/java/TestSpells.java]: comments to be deleted

File: src/main/java/org/example/Clazz.java (hunk 4 above)
Comment (agabylicka) [reply src/main/java/org/example/Clazz.java]: Done

File: src/main/java/org/example/Feature.java (hunk 5 above)
Comment (agabylicka) [reply src/main/java/org/example/Feature.java:16]: Done

File: src/main/java/org/example/RaceMAPPER.java (hunk 9 above)
Comment (agabylicka) [reply src/main/java/org/example/RaceMAPPER.java]: Done

File: src/main/java/org/example/race_to_parse.json
Comment (agabylicka) [reply src/main/java/org/example/race_to_parse.json:1]: Done

File: src/test/java/TestClass.java (hunk 17 above)
Comment (agabylicka) [reply src/test/java/TestClass.java]: Done

File: src/test/java/TestSpells.java (hunk 23 above)
Comment (agabylicka) [reply src/test/java/TestSpells.java]: Done

File: src/main/java/org/example/Character.java (hunk 2 above)
Comment (agabylicka) [reply src/main/java/org/example/Character.java:5]: Fixed

File: src/main/java/org/example/ClazzMAPPER.java (hunk 3 above)
Comment (agabylicka) [reply src/main/java/org/example/ClazzMAPPER.java]: Fixed

File: src/test/java/TestAbilityBonus.java (hunk 15 above)
Comment (agabylicka) [reply src/test/java/TestAbilityBonus.java:23]: Done

File: src/main/java/org/example/FeatureMAPPER.java (hunk 6 above)
Comment (agabylicka) [reply src/main/java/org/example/FeatureMAPPER.java]: Done

File: src/main/java/org/example/ItemMAPPER.java (hunk 7 above)
Comment (agabylicka) [reply src/main/java/org/example/ItemMAPPER.java]: Done

File: src/main/java/org/example/Language.java (hunk 8 above)
Comment (agabylicka) [reply src/main/java/org/example/Language.java:3]: Done

File: src/main/java/org/example/SpellsMAPPER.java (hunk 11 above)
Comment (agabylicka) [reply src/main/java/org/example/SpellsMAPPER.java]: Done

File: src/main/java/org/example/RaceMAPPER.java (hunk 10 above)
Comment (agabylicka) [reply src/main/java/org/example/RaceMAPPER.java:43]: Done

File: src/test/java/TestAbilityBonus.java (hunk 14 above)
Comment (agabylicka) [reply src/test/java/TestAbilityBonus.java:17]: Done

File: src/test/java/TestAbilityBonus.java (hunk 16 above)
Comment (agabylicka) [reply src/test/java/TestAbilityBonus.java]: Done

File: src/test/java/TestClass.java (hunk 20 above)
Comment (agabylicka) [reply src/test/java/TestClass.java:74]: Done

File: src/main/java/org/example/ItemMAPPER.java [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+
+    public static Item convertJSON(String JSON) {
+        try {
+            Item item = new Item();
+            JsonNode node = MAPPER.readTree(JSON);
+            item.setName((node.get("name").asText()));
+            item.setQuantity(node.get("quantity").asInt());
+            return item;
+        } catch (JsonProcessingException e) {
+            //throw new RuntimeException(e);
```
Comment (blazejprzyluski) [inline src/main/java/org/example/ItemMAPPER.java:18]: Why the commented exception?

**Pull request merged**